src/main/scala/mesosphere/marathon/core/launchqueue/impl/LaunchQueueActor.scala

```diff
@@ -211,9 +211,22 @@ private[impl] class LaunchQueueActor(
     suspendedLaunchersMessages += actorRef -> deferredMessages
   }
 
+  /**
+    * Fetch an existing actorRef, or create a new actor if the given instance is scheduled.
+    * We don't need an actor ref to handle update for non-scheduled instances.
+    */
+  private def actorRefFor(instance: Instance): Option[ActorRef] = {
+    launchers.get(instance.runSpecId).orElse {
+      if (instance.isScheduled) {
+        logger.info(s"No active taskLauncherActor for scheduled ${instance.instanceId}, will create one.")
+        groupManager.runSpec(instance.runSpecId).map(createAppTaskLauncher)
+      } else None
+    }
+  }
+
   private[this] def receiveInstanceUpdate: Receive = {
     case update: InstanceChange =>
-      launchers.get(update.runSpecId) match {
+      actorRefFor(update.instance) match {
```

Comment:
I don't understand. Why is this required? Why not stick to `update.runSpecId`?

Reply:
@alenkacz has the answer

Reply:
Well before, every terminal instance was expunged, so when we scheduled replacement, it always had to go through `LaunchQueue.Add` and that one created TaskLauncherActor. But now we do incarnations and don't expunge so for every instance update we might end up in a situation when an instance is scheduled and there has to be TaskLauncherActor for that instance. So this code change addresses exactly that

Reply:
Hm, it feels like we should rethink the `TaskLauncherActor` management. Eg what happens when we purge, ie stop the actor, and get an update? We would restart the actor even though it should have been gone, eg because of an app deletion.

Reply:
Agreed, the logic here is prone to races (and, from our reading, was race-prone before). Alena and I thought about it and we don't really see a way to resolve it. The probability of leaking TaskLauncherActors seems pretty low, and it's possible that we made it worse by this change. :/

Reply:
I have to think about this. I'm not even sure why the launch queue forwards the updates. The TaskLauncherActor could just subscribe. Well, maybe that would be too much of a change.